src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/CodeGeneration/MethodGenerator.cs

```diff
@@ -228,12 +227,24 @@ private static SyntaxList<TypeParameterConstraintClauseSyntax> GenerateDefaultCo
                 if (!referencedTypeParameters.Contains(typeParameter))
                     continue;
 
-                var constraint = typeParameter switch
+                TypeParameterConstraintSyntax? constraint;
+
+                if (typeParameter.HasReferenceTypeConstraint)
+                {
+                    constraint = s_classConstraint;
+                }
+                else if (typeParameter.HasValueTypeConstraint)
+                {
+                    constraint = s_structConstraint;
+                }
+                else if (typeParameter.ConstraintTypes.Any(t => t.IsReferenceType))
+                {
+                    constraint = s_classConstraint;
+                }
+                else
                 {
-                    { HasReferenceTypeConstraint: true } => s_classConstraint,
-                    { HasValueTypeConstraint: true } => s_structConstraint,
-                    _ => s_defaultConstraint
-                };
+                    constraint = s_defaultConstraint;
+                }
```

Comment:
i think this algorithm needs to be recursive to some extent (probably avoiding circularities).

Reply:
Why would it need to be recursive? Isn't the outermost type the one that will determine if it's a reference or value type? For example, there is no `T` that makes `ImmutableArray<T>` a reference type.

Reply:
Specifically, it *may* need to be recursive if you have `where T : U` and `where U : struct`.  @genlu do you have tests for these sorts of cases?  if so, and the above code works, *great*.  If no tests, please add them.  If added tests reveal this isn't working, please fixup code.

Thanks!